# Abnum 3

Abnum 3 is an alphabetic numerals library for Python 3. Library contains various letter value substituting systems from ancient times to modern artificial ones.

_Older version for Python 2 can be found from: https://github.com/markomanninen/abnum_

Currently supported scripts are: Greek (ancient), Hebrew and English. Planned for future is a support for Arabic, Coptic, Sanskrit, Aramean and Ethiopian scripts, which all are good candidates to have been used for alpha-numeric calculations in the past.

Some of the systems have been referred with certain name and methodology:

* Greek as [isopsephy](http://en.wikipedia.org/wiki/Isopsephy)
* Hebrew as [gematria](http://en.wikipedia.org/wiki/Gematria)
* Arabic as [abjad](http://en.wikipedia.org/wiki/Abjad_numerals)
* Sanskrit as [katapayadi](http://en.wikipedia.org/wiki/Katapayadi_system)

But not all languages and scripts have specific name for this exciting literary device. For that reason and purely for  recreational purposes I decided to put all systems under the same generic library, namely: Abnum 3.

Many languages have acrophonic (not alphabetic) numerals also. Most well known of them are roman numbers: I, V, X, L, C, D and M. Greeks had Attic (e.g. Herodianic) numerals, but any of these should not be confuced with alphabetic numerals implemented on the Abnum 3 library.

# Usage

Import Abnum class and scripts from the abnum package:

In [1]:
from abnum import Abnum, greek, hebrew, english

## Greek

Initialize abnum object with Greek script. Object (g) will hold all main functionality for alphabet numeral value calculation and transliteration purposes:

In [2]:
g = Abnum(greek)

Set up a word, or a phrase "o Logos" which means "the Word" for calculation:

In [3]:
phrase = 'ο Λογος'

You can transliterate Greek letters to Roman equivalents with convert method:

In [4]:
print (g.convert(phrase))

o Logos


Finally get the isopsephic value with value method of the Abnum class:

In [5]:
print (g.value(phrase))

443


Thus we can see that ο Λογος (the Word as a philosophical/theological concept) is transliterated to 'o Logos'. Its alphabetic numeral value (isopsephic value in the case of Greek example) is **443**.

Abnum uses [romanize](https://github.com/markomanninen/romanize) transliteration module for converting a script to Roman letters. But is also works other way around. You can get a Greek script presentation of the word/the phrase as well:

In [6]:
print (g.convert('O LOGOS'))

Ο ΛΟΓΟΣ


Some of the scripts have capital letters and some doesn't. This is affected on the functioning in a way that for example in Greek and English small and capital letters have same value and their form is reserved on conversion.

In [7]:
print (g.convert('GAIOS IOYLIOS KAISAR'), ',', \
       g.convert('GAIOS IOYLIOS KAISAR'.title()))

ΓΑΙΟΣ ΙΟΨΛΙΟΣ ΚΑΙΣΑΡ , Γαιοσ Ιοψλιοσ Καισαρ


Generally capitals doesn't have any effect on the value of the phrase. Thus "ΓΑΙΟΣ ΙΟΨΛΙΟΣ ΚΑΙΣΑΡ" and "Γαιοσ Ιοψλιοσ Καισαρ" should yield same value:

In [8]:
print (g.value('ΓΑΙΟΣ ΙΟΨΛΙΟΣ ΚΑΙΣΑΡ'), '=', g.value('Γαιοσ Ιοψλιοσ Καισαρ'), '=', g.value('gaios ioylios kaisar'))

1706 = 1706 = 1706


Yes, also Roman letters can be used for value calculation. But in this case they have Greek value substitution. For english "gematria" see example further below.


## Hebrew

Scripts like Hebrew doesn't have separate small or capital letters. Transliteration module in this case have special meaning on capital letters and their numerical value is not same with small letters. Capital "M" has a value 600 and small "m" 40 in the Hebrew gematria. Big letter are used as final letters as described by ancient [Mispar gadol](http://en.wikipedia.org/wiki/Gematria#Values_table) method of gematria. See the difference of spelling and value for the Hebrew word God:

In [9]:
h = Abnum(hebrew)
print (h.value('alhiM'), h.convert('alhiM'))
print (h.value('alhim'), h.convert('alhim'))

649 אלחים
89 אלחימ


Also note that reading direction of the Hebrew word letters are from right to left!

Let's try for a little bit longer phrase. Next one is from Hebrew version of the John's Revelation (chapter 1, verse 1). Preprocessing method is used to remove all diacritics which has no meaning on the value conversion. 

In [10]:
phrase = "גלינא דישׁוע משׁיחא דיהב לה אלהא למחויו לעבדוהי מא דיהיב למהוא בעגל ושׁודע כד שׁלח ביד מלאכה לעבדה יוחנן"
phrase = h.preprocess(phrase)
for word in phrase.split():
    print ("%s\t%s\t%s" % (word, h.convert(word), h.value(word)))
print ('\tTOTAL:\t%s' % h.value(phrase))

גלינא	clina	94
דישוע	disfo	390
משיחא	msiha	359
דיהב	dieb	21
לה	le	35
אלהא	alea	37
למחויו	lmhfif	100
לעבדוהי	lobdfei	127
מא	ma	41
דיהיב	dieib	31
למהוא	lmefa	82
בעגל	bocl	105
ושודע	fsfdo	386
כד	kd	24
שלח	slh	338
ביד	bid	16
מלאכה	mlake	96
לעבדה	lobde	111
יוחנן	ifhnN	774
	TOTAL:	3167


## English

Value conversion is made by default mapping, but in some cases (either because of using different letter value mapping from the various traditions or artificial one for tests) you may want to provide your own mapping. Take for example default english data mapping which is extensively used by [Marty Leeds](https://grahamhancock.com/leedsm1/):

In [11]:
e1 = Abnum(english)
for i in range(13):
    print (e1.data[:13][i][1], ',', e1.data[13:][i][1], '=', e1.data[:13][i][0])

a , n = 1
b , o = 2
c , p = 3
d , q = 4
e , r = 5
f , s = 6
g , t = 7
h , u = 6
i , v = 5
j , w = 4
k , x = 3
l , y = 2
m , z = 1


By using this default mapping next phrase "the key" would give a value accordingly:

In [12]:
word = "the key"
print ("Using a default english mapping and getting a value for the phrase: '%s' = %s" % (word, e1.value(word)))

Using a default english mapping and getting a value for the phrase: 'the key' = 28


But in english script, there are several [different mappings](http://www.gematrix.org/gematria.php#english_gematria), which you could use instead. So let's apply custom mapping instead and see a value for the phrase "the key":

In [13]:
mapping = ([1,  'a'], [2,  'b'], [3,  'c'], [4,  'd'], [5,  'e'], [6,  'f'], [7,  'g'], [8,  'h'], [9, 'i'], 
           [10, 'j'], [11, 'k'], [12, 'l'], [13, 'm'], [14, 'n'], [15, 'o'], [16, 'p'], [17, 'q'], [18, 'r'], 
           [19, 's'], [20, 't'], [21, 'u'], [22, 'v'], [23, 'w'], [24, 'x'], [25, 'y'], [26, 'z'])
e2 = Abnum(english, mapping)
print ("Using a custom  english mapping and getting a value for the phrase: '%s' = %s" % (word, e2.value(word)))

Using a custom  english mapping and getting a value for the phrase: 'the key' = 74


# Character tables

Abnum library offers some useful table visualization tools for "gematrians". First one is a char_table function. On the next example the famous Greek votive ἀδελφέ, βοήθει (Help, Brother!) is split to character specific representation of the values. Word are grouped as well with appropriate isopsephic value. At the end whole sentence value is shown.

In [14]:
g.char_table(g.preprocess('ἀδελφέ, βοήθει'))

<abnum.remarkuple.table.table.<locals>.table at 0x107c70dd8>

Same data can be viewed by pivotted table format:

In [15]:
a, b, c = g.char_table_data(g.preprocess('ἀδελφέ, βοήθει'))

In [16]:
a

letter transliteration  value    word
0       α               a      1  αδελφε
1       δ               d      4  αδελφε
2       ε               e      5  αδελφε
3       λ               l     30  αδελφε
4       φ               f    500  αδελφε
5       ε               e      5  αδελφε
6       β               b      2  βοηθει
7       ο               o     70  βοηθει
8       η               ê      8  βοηθει
9       θ               h      9  βοηθει
10      ε               e      5  βοηθει
11      ι               i     10  βοηθει

In [17]:
b

value  characters  digital_sum  digital_root
word                                                
αδελφε    545           6           14             5
βοηθει    104           6            5             5

In [18]:
c

characters  digital_root  digital_sum         phrase  value
0          12             1           19  αδελφε βοηθει    649

As you may notice, the isopsephy value of "Help, Brother" (αδελφε βοηθει) is same as the gematria value of "God" (אלחים). Drawing connections between values retrieved from different scripts should be done with the highest caution. Even the basic usage of the evaluations, that "the same value of different words makes words internally related", can be questioned many ways. But that is a different topic and out of the scope of the document.

# Advanced search functionality

Last, but not the least, Abnum 3 library supports finding number patterns from phrases with the _find_ -method. This is useful for examining possible word and phrase patterns on bigger text corpuses. This is actually the main reason why library was made.

As a real historical example I will use a poem created by first and second century isopsephist the architect Niko, (possibly the father of Claudius Galen), who was said to have mastered isopsephic art to its extremes.

![Neikoonos](neikoonos.jpeg)

Original plate (artifact available in the Archeological Museum of [Pergamon](https://en.wikipedia.org/wiki/Pergamon), Turkey) has second part of the poem broken, but the first part is fully reconstructed and translated:

_"Helios, you allow the light to assume its spiral shape with your <br/>
swith horses, thus you have sent every day your rays to men having <br/>
completed everything, since you established the route of Sun and <br/>
(then) the never ending Earth, the outpuring Water, Air and Fire, <br/>
everything proceeds ordinally."_

Let's Abnum library calculate the value of the each line (in Greek) and then find number patterns that were in the mind of original author.

In [19]:
ss = ["ΑΙΛΙΟΥ ΝΕΙΚΩΝΟΣ ・ ẠΨΚϚ ・ ΑΡΧΙΤΕΚΤΟΝΟΣ",
      "ΗΛΙΕ ΘΟΑΙΣ ΙΠΠΟΙΣΙΝ ΕΙΛΙΣΣΩΝ ΦΛΟΓΑ",
      "ΩΣ ΠΑΝΤΕΛΗ ΘΝΗΤΟΙΣΙ ΤΗ ΤΟΤΕ Γ ΗΜΕΡΑ",
      "ΑΚΤΕΙΝΑΣ ΕΦΗΚΑΣ ΘΕΜΕΝΟΣ ΗΛΙΟΥ ΔΡΟΜΟΥΣ",
      "ΚΑΙ ΤΗΝ ΑΠΕΙΡΟΝ ΓΑΙΑΝ ΗΔΕ ΥΓΡΟΥ ΧΥΣΕΙΣ",
      "ΑΕΡΑ ΤΕ ΚΑΙ ΠΥΡ ΕΝ ΤΑΞΕΙ ΦΟΡΟΥΜΕΝΑ",
      "ṂĖ"]

print ()
for s in ss:
    print (g.value(g.preprocess(s)), s)

s = g.preprocess(' '.join(ss))

print ()
print ("Total value: %s" % g.value(s))


5178 ΑΙΛΙΟΥ ΝΕΙΚΩΝΟΣ ・ ẠΨΚϚ ・ ΑΡΧΙΤΕΚΤΟΝΟΣ
2762 ΗΛΙΕ ΘΟΑΙΣ ΙΠΠΟΙΣΙΝ ΕΙΛΙΣΣΩΝ ΦΛΟΓΑ
3271 ΩΣ ΠΑΝΤΕΛΗ ΘΝΗΤΟΙΣΙ ΤΗ ΤΟΤΕ Γ ΗΜΕΡΑ
3102 ΑΚΤΕΙΝΑΣ ΕΦΗΚΑΣ ΘΕΜΕΝΟΣ ΗΛΙΟΥ ΔΡΟΜΟΥΣ
3175 ΚΑΙ ΤΗΝ ΑΠΕΙΡΟΝ ΓΑΙΑΝ ΗΔΕ ΥΓΡΟΥ ΧΥΣΕΙΣ
2690 ΑΕΡΑ ΤΕ ΚΑΙ ΠΥΡ ΕΝ ΤΑΞΕΙ ΦΟΡΟΥΜΕΝΑ
15000 ṂĖ

Total value: 35178


Then find phrases with a value 1726 and 15000:

In [20]:
v1 = 1726
v2 = 15000
print ()
for x in g.find(s, v1, True):
    print (v1, x)
print ()
for x in g.find(s, v2, True):
    print (v2, x)


1726 ΑΙΛΙΟΥ ΝΕΙΚΩΝΟΣ
1726 ẠΨΚϚ
1726 ΑΡΧΙΤΕΚΤΟΝΟΣ

15000 ΗΛΙΕ ΘΟΑΙΣ ΙΠΠΟΙΣΙΝ ΕΙΛΙΣΣΩΝ ΦΛΟΓΑ ΩΣ ΠΑΝΤΕΛΗ ΘΝΗΤΟΙΣΙ ΤΗ ΤΟΤΕ Γ ΗΜΕΡΑ ΑΚΤΕΙΝΑΣ ΕΦΗΚΑΣ ΘΕΜΕΝΟΣ ΗΛΙΟΥ ΔΡΟΜΟΥΣ ΚΑΙ ΤΗΝ ΑΠΕΙΡΟΝ ΓΑΙΑΝ ΗΔΕ ΥΓΡΟΥ ΧΥΣΕΙΣ ΑΕΡΑ ΤΕ ΚΑΙ ΠΥΡ ΕΝ ΤΑΞΕΙ ΦΟΡΟΥΜΕΝΑ
15000 ṂĖ


Finally you might want to examine individual word values. Char_table_data -function provides count of the characters plus digital sum and digital root of the word value as well.

In [21]:
a, b, c = g.char_table_data(s)
b

value  characters  digital_sum  digital_root
word                                                      
ΑΕΡΑ            107           4            8             8
ΑΙΛΙΟΥ          521           6            8             8
ΑΚΤΕΙΝΑΣ        587           8           20             2
ΑΠΕΙΡΟΝ         316           7           10             1
ΑΡΧΙΤΕΚΤΟΝΟΣ   1726          12           16             7
Γ                 3           1            3             3
ΓΑΙΑΝ            65           5           11             2
ΔΡΟΜΟΥΣ         884           7           20             2
ΕΙΛΙΣΣΩΝ       1305           8            9             9
ΕΝ               55           2           10             1
ΕΦΗΚΑΣ          734           6           14             5
ΗΔΕ              17           3            8             8
ΗΛΙΕ             53           4            8             8
ΗΛΙΟΥ           518           5           14             5
ΗΜΕΡΑ           154           5           10             1
ΘΕΜΕΝΟΣ         379           7           19             1
ΘΝΗΤΟΙΣΙ        657           8           18             9
ΘΟΑΙΣ           290           5           11             2
ΙΠΠΟΙΣΙΝ        510           8            6             6
ΚΑΙ              62           6            8             8
ΝΕΙΚΩΝΟΣ       1205           8            8             8
ΠΑΝΤΕΛΗ         474           7           15             6
ΠΥΡ             580           3           13             4
ΤΑΞΕΙ           376           5           16             7
ΤΕ              305           2            8             8
ΤΗ              308           2           11             2
ΤΗΝ             358           3           16             7
ΤΟΤΕ            675           4           18             9
ΥΓΡΟΥ           973           5           19             1
ΦΛΟΓΑ           604           5           10             1
ΦΟΡΟΥΜΕΝΑ      1236           9           12             3
ΧΥΣΕΙΣ         1415           6           11             2
ΩΣ             1000           2            1             1
ṂĖ            15000           2            6             6
ẠΨΚϚ           1726           4           16             7

At least count of the characters have played big role on the text composition process, maybe even bigger than isopsephic value calculation. If text corpus was supposed to be copied, then counting characters served at least two purposes: a) to estimate cost of the copying and b) providing information of the accuracy of the copying process. But what comes to isopsephy/gematria, accuracy of the copying process could be even further rechecked if value patterns of the text was cautiously selected and applied to the text corpus.

In [22]:
from IPython.core.display import HTML
with open('abnum.css') as f:
    css = f.read()
HTML('<style>%s</style>' % css)

## The [MIT](http://choosealicense.com/licenses/mit/) License

Copyright &copy; 2016 Marko Manninen